blog/2023-07-24-pwa-tech-1.md/index.md

```diff
@@ -0,0 +1,474 @@
+
+These do not cover _all_ of the static assets in the app’s `build` directory however; other files like icons, web manifests, and javascript files from vendors like `jQuery` need to be handled separately. To add those remaining files to the precache manifest, we added another step to _our_ CLI’s build process. After executing the CRA build step, we use the [`injectManifest`](https://developers.google.com/web/tools/workbox/modules/workbox-build#injectmanifest_mode) function from the [`workbox-build`](https://developers.google.com/web/tools/workbox/modules/workbox-build) package to read all of the _other_ static files in the app’s `build` directory, generate a manifest of those URLs, and inject _that_ list into the compiled service worker at a prepared placeholder. You can see the resulting `injectManifest` code [here](https://github.com/dhis2/app-platform/blob/master/cli/src/lib/pwa/injectPrecacheManifest.js).
+
+Handling these precache manifests correctly is also important for keeping the app up-to-date, which will be described in the [“Managing the service worker’s updates and lifecycle” section](#managing-the-service-workers-updates-and-lifecycle) below.
+
+### Using a config option to enable PWA features
+
+To implement the opt-in nature of the PWA features, the service worker should only be registered if PWA is enabled in the app’s [configuration](https://platform.dhis2.nu/#/config). We added an option to the [`d2.config.js` app config file](https://platform.dhis2.nu/#/config/d2-config-js-reference) that can enable PWA, which looks like this:
+
+```diff title="d2.config.js"
+module.exports = {
+    type: 'app',
+    title: 'My App',
+
++   pwa: { enabled: true },
+
+    entryPoints: {
+        app: './src/App.js',
+    },
+}
+```
+
+During the `d2-app-scripts` `start` or `build` processes, the config file is read, and a `PWA_ENABLED` value is added to the app’s environment variables. Then, in the App Adapter’s initialization logic, it registers or unregisters the service worker based on the the `PWA_ENABLED` environment variable.
+
+The registration logic will be described in more detail in the ["Registration of the service worker"](#registration-of-the-service-worker) section below.
+
+### Managing the service worker’s updates and lifecycle
+
+Managing the service worker’s [lifecycle](https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle) is both complex and vitally important. Because the core assets that run the app are precached and served directly from the cache without contacting the server, in order for a client to use newly deployed app updates, the service worker must be updated with a new precache manifest.
```

Comment:
Good call that this can be improved -- I think the causal link between the things is important (i.e. "because"), but I'll try to rewrite this to be clearer

Reply:
It took me a couple tries to rewrite this in a concise way that also conveys the meaning that I wanted -- what do you think of this?
After "Managing the service worker's lifecycle is both complex and vitally important":
> Because the service worker is responsible for serving the app from cached files, it now has a role in what version of the app a user sees. Note that the service worker serves the app from a list of files that’s set at the time it gets compiled. Because of this, the service worker _itself_ needs to be updated in a user’s browser in order to serve an updated version of the app.

Reply:
Thats seems very clear!